Update the resource revisions guide using nested collections (take 2) (#1192)
googleapis/aip: aip/general/0162.md

## Guidance

APIs **may** store a revision history for a resource. Examples of when it is
useful include:

- When it is valuable to expose older versions of a resource via an API. This
  can avoid the overhead of the customers having to write their own API to store
  and enable retrieval of revisions.
- Other resources depend on different revisions of a resource.
- There is a need to represent the change of a resource over time.

APIs implementing resources with a revision history **should** abstract resource
revisions as nested collection of the resource. Sometimes, the revisions
collection can be a top level collection, exceptions include:

- If resource revisions are meant to have longer lifespan than the parent
resource. In other words, resource revisions exist after resource deletion.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;

   // The snapshot of the book
   Book snapshot = 2;

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // Other revision IDs that share the same snapshot.

>   repeated string alternate_ids = 4

Reviewer: Should we rename this to `aliases`? (That would be more consistent with other changes.)

Author: agreed, done.

## Guidance

APIs **may** store a revision history for a resource. Examples of when it is
useful include:

- When it is valuable to expose older versions of a resource via an API. This
  can avoid the overhead of the customers having to write their own API to store
  and enable retrieval of revisions.
- Other resources depend on different revisions of a resource.

> - There is a need to visualize the change of a resource over time.

Reviewer 2: Nit: just think "visualize" is the wrong word 🤷🏻 
```suggestion
- There is a need to represent the change of a resource over time.
```

APIs implementing resources with a revision history **should** abstract resource
revisions as nested collection of the resource. Sometimes, the revisions
collection can be a top level collection, exceptions include:

- If resource revisions are meant to have longer lifespan than the parent

> resource. In other words, resource revisions exist after the resource deletion.

Reviewer 2: Nit
```suggestion
resource. In other words, resource revisions exist after resource deletion.
```

> The resource revision type **must** have a field named `snapshot` of the parent

Reviewer 2: This feels a tad out of place, like it belongs with the bullets under the example. In fact, it looks like a dupe of a bullet already there.

Author: agreed, I think it was the natural evolution based on the comments. moved.

resource type.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;
   
   // The snapshot of the book
   Book snapshot = 2;

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // Other revision IDs that share the same snapshot.
  repeated string alternate_ids = 4
    [(google.api.field_behavior) = OUTPUT_ONLY];
}
```

> - The resource revision **must** contain a field with a message type of the

Reviewer 2: Didn't see these mentioned explicitly anywhere.
```suggestion
- The `message` name **must** be named like `{ResourceType}Revision`.
- The `message` **must** be annotated as a resource (AIP-123).
- The resource revision **must** contain a field with a message type of the
```

Author: applied!

  parent resource, with a field name of the resource singular.
- The resource revision **must** contain a `create_time` field (see [AIP-142][]).
- The resource revision **may** contain a repeated field `alternate_ids`, which would
  contain a list of resource IDs that the revision is also known by (e.g. `latest`)

### Creating Revisions

Depending on the resource, different APIs may have different strategies for
when to create a new revision, such as:

> - Create a new revision any time that there is a change

Reviewer 2: Change to what? Maybe:
```suggestion
- Create a new revision any time that there is a change to the parent resource
```

Author: these are lifted from the previous version I think, but definitely makes sense to fix up wording. Updated.

> - Create a new revision when something important happens

Reviewer 2: Nit: "something important happens" isn't concrete enough for me, but that's just me.
```suggestion
- Create a new revision when important system state changes
```

> - Create a new revision when the user specifically asks

Reviewer 2: Nit:
```suggestion
- Create a new revision when specifically requested
```

APIs **may** use any of these strategies. APIs **must**
document their revision creation strategy.

### Resource names for revisions

When referring to specific revision of a resource, the subcollection name
**must** be named `revisions`. Resource revisions have names with the format
`{resource_name}/revisions/{revision_id}`. For example:
```
publishers/123/books/les-miserables/revisions/c7cfa2a8
```

### Server specified IDs

> Services **may** reserve specific IDs to be server-specified (e.g. `current`,

Reviewer 2: These aren't IDs, these are aliases, no? https://google.aip.dev/122#resource-id-aliases

Author: oh nice! I didn't realize it's common nomenclature. works great.

Author: updated, PTAL.

## Guidance

APIs **may** store a revision history for a resource. Examples of when it is
useful include:

- When it is valuable to expose older versions of a resource via an API. This
  can avoid the overhead of the customers having to write their own API to store
  and enable retrieval of revisions.
- Other resources depend on different revisions of a resource.
- There is a need to visualize the change of a resource over time.

APIs implementing resources with a revision history **should** abstract resource
revisions as nested collection of the resource. Sometimes, the revisions

> collection can be a top level collection, exception includes:

Reviewer: exceptions include?

- If resource revisions are meant to have longer lifespan than the parent
resource. In other words, resource revisions exist after the resource deletion.

> The resource revision type

Reviewer: Nit: rewrap the text.

But more importantly, I don't think "inherit" is a term we use in AIPs. AIP-122 mentions it, but only in passing.

> The resource revision type **must** inherit from the parent resource and has a

Reviewer: We've still got "inherit" here, which I don't think is helpful without elaboration.

Author: strange, I thought I updated this.

I'll merge and ensure it's fixed before resolving again.

field being the parent resource type.

```proto
message BookRevision {
   // The name of the book revision.
   string name = 1;

   // The snapshot of the book
   Book book = 2;

  // The timestamp that the revision was created.
  google.protobuf.Timestamp create_time = 3
    [(google.api.field_behavior) = OUTPUT_ONLY];

  // Other revision IDs that share the same snapshot.

>   repeated string alternate_ids = 4

Reviewer: So if I call GetBookRevision with a name of "books/xyz/revisions/latest" or a tag, what should happen? Should `name` be the name I requested, and the canonical name is returned in alternate_ids, or is the same representation returned in all cases, with the name being the canonical name, and alternate_ids including "latest" and any tags created? I suspect the latter is more useful, but I don't know whether it *could* confuse Terraform etc.

Author: Terraform likely won't use revisions, but definitely good to consider it. 

I think it has to be the former: returning the resource name the user sent. Otherwise, it's deviating behavior specified in the standard method AIPs.

One of my personal goals here is to remove exceptional cases, especially in the standard methods: if we introduce scenarios where a GET of a resource with an ID foo returns back a resource with an ID bar, that's another potential behavior exception that clients must follow. WDYT?

Reviewer: It feels to me like the latter is closer to https://google.aip.dev/122#resource-id-aliases:

> APIs may provide programmatic aliases for common lookup patterns. However, all data returned from the API must use the canonical resource name.

Is the "always return what the user asked for" specified anywhere, other than for project name/ID in AIP-2510?

Author: For aliases specifically, there's this line:

```
APIs may provide programmatic aliases for common lookup patterns. However, all data returned from the API must use the canonical resource name.
```

which I suppose implies that aliases should not use the resource name sent.

Author: given that this is documented in 122, should I add the clause here?

Author: I'll leave it in 122 for now (since duplication often leads to drift).

  repeated string alternate_ids = 4;

>   [(google.api.field_behavior) = OUTPUT_ONLY];

Reviewer: Nit: indent?

}
```

- The resource revision **must** contain a field with a message type of the
  parent resource, with a field name of the resource singular.
- The resource revision **must** contain a `create_time` field (see [AIP-142][]).

> - The resource revision **may** contain a repeated field `tags`, which would

Reviewer: Is this `alternate_ids` above? (I can't tell whether we're moving away from the term "tag" to have "alternate_id" or vice versa.)

Author: sorry, you're right. the goal is to move away from `tags` and just refer to everything as resource IDs. then tags just become a convenience custom method.

Author: updated to use the correct term `alternate_ids`.

  contain a list of resource IDs that the revision is also known by (e.g. `current`)

### Creating Revisions

Depending on the resource, different APIs may have different strategies for
when to create a new revision, such as:

- Create a new revision any time that there is a change
- Create a new revision when something important happens
- Create a new revision when the user specifically asks

> APIs **may** use any of these strategies.

Reviewer: Possibly add:

APIs **must** document their revision creation strategy.

?

Author: added. It'll be difficult to enforce, but from a best practices standpoint agreed on must.

> APIs **may** use any of these strategies. If a strategy is used, they **must**

Reviewer: I'd remove "if a strategy is used" - there must be *a* strategy, otherwise there'll never be any revisions :)

Author: ack, fixed.

### Resource names for revisions

When referring to specific revision of a resource, the subcollection name

> **must** named `revisions`. Resource revisions have names with the format

Reviewer: Nit: missing "be"

> `{resource_name}/revisions/{revision_id}`. For example:

Reviewer: Do we need to repeat the part about "this is only a should there are exceptions where you might want to make it a top-level collection"? (And should we give guidance about the name in that case?)

Author: more guidance is always helpful: is that clear? if not, I think we can always follow-up on this PR.

Author: I kind of imagined 162 being specific to subcollection-style revisions, without guidance for non-subcollection resources and that being out of scope of this guidance.

But if you feel differently happy to incorprate something precise.

Reviewer: I suspect most of the guidance still applies anyway.. but maybe the right thing to do is add something around line 46 saying "The rest of this AIP assumes a revisions collection as a sub-collection of the main resource collection" or similar.

Author: I think this is do-able... but will probably be a follow-up draft PR, since I doubt I can flesh out a well-defined enough AIP in time.

Author: I'll resolve this for now, since we'll target sibling collections in a future PR.

> publishers/123/books/les-miserables/revisions/c7cfa2a8

Reviewer: I suspect this should be in backticks.

Author: there shouldn't be a reference to current anyway, removed.

Author: sorry my editor put this comment in a weird place. wrapped in a code block.

### Server specified IDs

Services **may** reserve specific IDs to be server-specified (e.g. `current`,
`latest`), which are read-only and managed by the service.

```
GET /v1/publishers/{publisher}/books/{book}/revisions/{revision_id}
```

> - If a `latest` ID exists, it **must** represent the most recently created

Reviewer: Possibly explain how "latest" and "current" are normally expected to differ?

Author: yeah, let me seee if there was naything in the original design doc around this. I suspect we never defined the two particularly well.

maybe just remove `current` altogether? I think it might be more confusing than less.

Author: I've just removed the mention of a `current` ID, since `latest` is the only well defined one. The old version didn't have a concept of `current` ID anyway.

Reviewer: Cool. I'll have another look through all the changes tomorrow.

revision. The content of `publishers/{publisher}/books/{book}/revisions/latest`
and `publishers/{publisher}/books/{book}` can differ, as the latest revision may
be different from the current state of the resource.

### Tagging

> Rather than create a revision via the standard method, it may be more convenient

Reviewer: It feels like we're currently in a bit of a half-way house between referring to tags and not doing so. At the moment we have both concepts not just in the description, but in code, via the `alternate_ids` field and the `TagXyzRevision` custom method.

I don't think this is blocking, but I always get nervous when we refer to basically the same thing with two different terms - that's tended to work out badly over time, in my experience. If we can justify it, that's fine.

Author: I do agree with nomenclature discrepancies (alternate_ids vs tags). 

The way I was looking at this was "tagging" is an action that you can do with revisions, but there is no "tag" object: it's just a custom method for convenience.

Do you think that's not clear from this AIP? or maybe a different suggestion?

ack on this being non-blocking, just always a good time to have the discussion while we both have context.

Reviewer: This is more about terminology than actual resources.

If we want to use tag/tagging only as a verb, this should be rewritten:

> This allows users to write code against specific tags (e.g. published) and the revision can change in the background with no code change.

We might also want to make it really clear in the "Using resource ID instead of tag" - explicitly say that this is what `alternate_ids` (or `aliases` if we rename it) is for.

Author: resolved in the api design sync, we'll use aliases instead.

to create a revision snapshotting the current resource configuration.

APIs **may** provide a mechanism for users to use the existing revisions or
current resource state to create a new revision with a custom method "tag":

```proto
rpc TagBookRevision(TagBookRevisionRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*}/revisions:tag"
    body: "*"
  };
}
```

```proto
message TagBookRevisionRequest {
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/Book"
    }];

>   // If set, the specified id will be used. Otherwise, a random

Reviewer: Nit: ID instead of id. (We use ID in the next sentence.)

  // ID will be provided.
  string revision_id = 2 [(google.api.field_behavior) = OPTIONAL];

  // If unset, the current resource configuration will used to create the
  // revision.
  //
  // If set, the configuration in the specified revision will be used. This is
  // helpful when one wants to set an alias for the resource, such as

>   // `CURRENT`.

Reviewer: Use "current" instead of "CURRENT" here, to be consistent with the earlier guidance?

Author: agreed, fixed.

  string source_revision_id = 3 [(google.api.field_behavior) = OPTIONAL];
}
```

- The request message **must** have a `name` field:
  - The field **must** be [annotated as required][aip-203].
  - The field **must** identify the [resource type][aip-123] that it
    references.
- The request message **must** have a `revision_id` field:
  - The field **must** be [annotated as optional][aip-203].
- The request message **must** have a `source_revision_id` field:
  - The field **must** be [annotated as optional][aip-203].

> - If the user calls the `Tag` method with an existing revision_id, the request

Reviewer: Possibly just "the tag method" given that there isn't a method with the actual name `Tag`?
(And put `revision_id` in backticks.)

Author: agreed, fixed.

  **must** succeed and the revision will be updated with the configuration for
  the specified resource. This allows users to write code against specific tags
  (e.g. `published`) and the revision can change in the background with no code

>   change.

Reviewer: Doesn't that mean that designating the resource field IMMUTABLE is a bit disingenous? I'd expect that to mean "If I fetch the same URL twice, I should see the same content in that field both times...

Author: I think you're right. I've removed the `IMMUTABLE` field behavior.

### Rollback

A common use case for a resource with a revision history is the ability to roll
back to a given revision. APIs **should** handle this with a `Rollback` custom
method:

```proto
rpc RollbackBook(RollbackBookRequest) returns (BookRevision) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*}:rollback"
    body: "*"
  };
}
```

- The method **must** use the `POST` HTTP verb.
- The method **should** return a resource revision.

```proto
message RollbackBookRequest {
  // The book being rolled back.
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];
}
```

- The request message **must** have a `name` field to identify the resource

>   revision being rolled back.

Reviewer: I think this is unclear at the moment. I suspect the intention is for it to identify the resource revision which the resource is being rolled back *to*. In other words, it identifies what the resource should look like *afterwards* - is that right?

How do the pseudo-revision-IDs e.g. "current" and "latest" fit in here?

Author: yeah, you're right, fixed.

I don't think we have to mention `current` and `latest` explicitly here: the other section basically just says they're server-managed special resource IDs, so `name` doesn't need to handle them any different than any othe resource IDs. WDYT?

Reviewer: Yup, that's much clearer. I think it's fine now.

publishers/123/books/les-miserables/revisions/c7cfa2a8
```

### Server-specified Aliases

> Services **may** reserve specific IDs to be an [alias][] (e.g.

Reviewer: I'd put it back to aliases, but link to alias: `[aliases][alias]`. Otherwise the plural/singular doesn't match.
We can fix that up afterwards though.

Author: fine to fix now! done.

and `publishers/{publisher}/books/{book}` can differ, as the latest revision may
be different from the current state of the resource.

### User-Specified Aliases

APIs **may** provide a mechanism for users to assign an [alias][] ID to an
existing revision with a custom method "alias":

```proto
rpc AliasBookRevision(TagBookRevisionRequest) returns (Book) {
  option (google.api.http) = {
    post: "/v1/{name=publishers/*/books/*/revisions/*}:alias"
    body: "*"
  };
}
```

```proto
message AliasBookRevisionRequest {
  string name = 1 [
    (google.api.field_behavior) = REQUIRED,
    (google.api.resource_reference) = {
      type: "library.googleapis.com/BookRevision"
    }];

  // The ID of the revision to alias to, e.g. `CURRENT` or a semantic
  // version.

>   string alias_id = 2 [(google.api.field_behavior) = REQUIRED];

Reviewer: If this is required, then there's no way of saying "Please take the latest state of the resource (even if there isn't already a specific revision for it) and create an alias revision." I suspect that would be a fairly common use case. Would you want that to be handled via CreateBookRevision instead? (If so, that means an alias can be specified there.)

We also still need to work out how to effectively delete aliases.

Author: I think we can't really do it in the `Create` call without having some sort of initial alias in the resource state, which might get stale if a different version creates itself. 

So I guess making this optional is the easiest way to do so? wand we can document the behavior as such.

For delete, I think we can add a clause that if the delete is an alias, then the alias is removed but the actual revision is not? It's a little wonky since the delete call wouldn't actually delete a resource, but in practice I can't think of any issues.

Reviewer: Delete: yes, that works for me.

Create: Aargh - I've confused things by adding my comment to `alias_id` instead of `name`. `alias_id` should be required, and `name` has to be required really, because otherwise we can't identify the original resource. It's almost like we want the field to refer to "the revision or the resource" ;)
Back when the aliasing operation was on the resource itself, this was okay... but now it's not :( Basically we want three pieces of information:

- The Book resource (required)
- The "target" alias ID (required)
- The "source" revision ID (optional)

And we've only got two fields. Darn.

Maybe we can just live with this for now.

Author: Yeah let me think about it - I recognize the issue too. 

How about we merge as-is, and I'll try to propose a solution to the "create a revision and alias" problem in a follow-up?

I don't think this is worse that the current revisions AIP - it's essentially the same workflow, so at least we're not moving backwards.

I will re-update this to be required though.

Reviewer: Yup, that works for me.

```proto
message RollbackBookRequest {

>   // The book revision being rolled back.

Reviewer: How about "The revision that the book should be rolled back to"?

Author: done!

- Child resources where each child resource is a child of _a single revision
  of_ the parent resource.

APIs **should not** include multiple levels of resources with revisions, as
this quickly becomes difficult to reason about.

### Standard methods

Any standard methods **must** implement the corresponding AIPs ([AIP-131][],

> [AIP-132][], [AIP-133][], [AIP-134][], [AIP-135][]).

Reviewer 3: The two links for [API-133] and [API-134] aren't working --- they need their footnote-style links at the bottom of the file.

Author: Actually we don't need hard-coded links to AIPs because the site-generator adds them for you.

So I'll just make this naked text to make it less error-prone.

  of_ the parent resource.

APIs **should not** include multiple levels of resources with revisions, as
this quickly becomes difficult to reason about.

### Standard methods

Any standard methods **must** implement the corresponding AIPs ([AIP-131][],
[AIP-132][], [AIP-133][], [AIP-134][], [AIP-135][]).

- List methods: By default, revisions in the list response **must** be ordered
  in reverse chronological order. User can supply `order_by` to override the
  default behavior.

As revisions are nested under the resource, also see [cascading delete][].

## Rationale

### Abstract revisions as nested collection

Revisions being resources under nested collection make revisions a first class
citizen.

- Revisions can offer standard get, list, and delete methods.
- It retains the flexibility of extending new fields to revision in addition to
  the resource message.

### Consolidation of tagging use cases

Previously, committing and tagging a resource were two separate methods on a
resource, enabling the following user journeys:

- Commit
  - Creating a revision with the current resource configuration.
- Tag
  - Taking an existing revision and giving it an alias.

> Tagging is effectively giving a second resource ID to a revision.

Reviewer 2: Maybe?
```suggestion
Tagging is effectively giving an alias to a revision.
```

Author: Good call, thanks!

- Child resources where each child resource is a child of _a single revision
  of_ the parent resource.

APIs **should not** include multiple levels of resources with revisions, as
this quickly becomes difficult to reason about.

### Standard methods

> If a revision implements standard methods, they **must** implement the standard

Reviewer: Is this intended to mean "any standard methods provided must implement the corresponding AIPs"? It's not intended to mean "if you implement some of these you must implement all of them" or anything like that? Happy to try to come up with a clarifying rewording if the former is the intent.

Author: the intent was the former. Great wording! added.

methods ([AIP-131][], [AIP-132][], [AIP-133][], [AIP-134][], [AIP-135][]).

- List methods: By default, revisions in the list response **must** be ordered
  in reverse chronological order. User can supply `order_by` to override the
  default behavior.

As revisions are nested under the resource, also see [cascading delete][].

## Rationale

### Abstract revisions as nested collection

Revisions being resources under nested collection make revisions a first class
citizen.

- Revisions can offer standard get, list, and delete methods.
- It retains the flexibility of extending new fields to revision in addition to
  the resource message.

### Consolidation of tagging use cases

Previously, committing and tagging a resource were two separate methods on a
resource, enabling the following user journeys:

- Commit
  - Creating a revision with the current resource configuration.
- Tag
  - Taking an existing revision and giving it an alias.

Tagging is effectively giving a second resource ID to a revision.

## History

### Switching from from a collection extension to a subcollection

> In 2023-09, revisions are abstracted as nested resource. Prior to this,

Reviewer: nested resource => nested resources, or "a nested resource collection" perhaps.

revisions are more like extension of an existing resource by using `@` symbol.

> List and delete revisions are custom method on the resource collection. Getting

Reviewer: custom method => custom methods

(And possibly "Operations to list and delete revisions"? (Getting really nitpicky here...)

> a revision shares the same method with get resource.

Reviewer: Sorry, I should really have read the second half of the sentence when nitpicking about the first. I don't know what's meant by "shares the same method with get resource".

Author: How does this look? I was trying to say Get was overloaded by retrieving either resource revisions, or the resource.

> It's primary advantage was allowing a resource reference to seamlessly refer to

Reviewer: It's => Its, here and in the next line. (Maybe check for "its" and "it's" throughout...)

Author: ack, looked through the doc and I think that's it. thanks!

a resource, or it's revision.

It also had several disadvantages:

-  List revisions is a custom method (:listRevisions) on the resource collection
-  Delete revision is a custom method on the resource collection
-  Not visible in API discovery doc
-  Resource ID cannot use `@`

The guidance was modified ultimately to enable revisions to behave like a
resource, which reduces the users cognitive load and allows resource-oriented
clients to easily list, get, create, and update revisions.

### Using resource ID instead of tag

> In th previous design, revisions had a separate identifer for a revision known

Reviewer: th => the

- Child resources where each child resource is a child of _a single revision
  of_ the parent resource.

APIs **should not** include multiple levels of resources with revisions, as
this quickly becomes difficult to reason about.

### Standard methods

If a revision implements standard methods, they **must** implement the standard
methods ([AIP-131][], [AIP-132][], [AIP-133][], [AIP-134][], [AIP-135][]).

- List methods: By default, revisions in the list response **must** be ordered
  in reverse chronological order. User can supply `order_by` to override the
  default behavior.

As revisions are nested under the resource, also see [cascading delete][].

## Rationale

### Abstract revisions as nested collection

Revisions being resources under nested collection make revisions a first class
citizen.

- Revisions can offer standard get, list, and delete methods.

> - Revision has its own separate message type and inherits from the parent

Reviewer: Again, I'd be nervous of referring to inheritance if we haven't really defined it.

Author: yeah, make sense. fixed up the wording here and removed the term inheritance, PTAL.

  resource type. It retains the flexibility of extending new fields to revision
  resources.

### Consolidation of tagging use cases

Previously, committing and tagging a resource were two separate methods on a
resource, enabling the following user journeys:

- Commit
  - Creating a revision with the current resource configuration.
- Tag
  - Taking an existing revision and giving it an alias.

Tagging is effectively giving a second resource ID to a revision.

## History

### Switching from from a collection extension to a subcollection

> In 2023-07, revisions are abstracted as nested resource. Prior to this,

Reviewer: Update to 2023-09? (Anticipating that we won't merge it for another couple of days at least)

Author: ack, done.

revisions are more like extension of an existing resource by using `@` symbol.
List and delete revisions are custom method on the resource collection. Getting

> revision shares the same method with get resource.

Reviewer: "Getting a revision"?

Author: fixed, thanks!

It had several advantages:

> - A resource reference could seamlessly refer to a resource, or it's revision.

Reviewer: it's => its

And perhaps if there's only one advantage we're going to list, it's at least worth removing "several"?

(We really need to do work on this resource-or-revision flexibility in the future...)

Author: fixed. I don't know if we can address the resource-or-revision fliexibility in this PR specifically... feel free to open if you think it's blocking, I don't know how prevalent the usage is, but I imagine a revision and the actual resource are worlds apart in terms of how the control and data plane treat them.

It also had several disadvantages:

-  List revisions is a custom method (:listRevisions) on the resource collection
-  Delete revision is a custom method on the resource collection

> -  Not visible in api discovery doc

Reviewer: api => API?

Author: fixed.

-  Resource ID cannot use `@`

The guidance was modified ultimately to enable revisions to behave like a
resource, which reduces the users cognitive load and allows resource-oriented
clients to easily list, get, create, and update revisions.

### Using resource ID instead of tag

> In, revisions had a separate identifer for a revision known as a `tag`,

Reviewer: "In," ?

("In the previous design", perhaps?)
